Hacktoberfest : Move documentation to Github (#8)
jenkinsci/folder-properties-plugin: README.md

The folder-properties plugin is a Jenkins plugin which allows users with config permission to define
properties for a folder which can then be used by any jobs contained
within it or in any of its sub-folders. 

# About this Plugin

The aim here is to remove the need to specify the same properties over
and over again for all the jobs inside a folder.

## How to use?

To configure, just create a
[folder](https://plugins.jenkins.io/cloudbees-folder/),
go to its configuration page and add as many properties as you need
under the `Folder Properties` section.

In structures where two or more folders are nested, any property defined for a folder will be overridden by any other
property of the same name defined by one of its sub-folders.

![](docs/images/folder-properties-config.png)

## Freestyle Jobs

Freestyle jobs must opt into the `Folder Properties` build wrapper from
the `Build Environment` section of their configuration page in order to

> be able to access these properties as they would any other environment

Reviewer: ```suggestion
be able to access these properties as they would any other environment
```
`as they would any other environment` seems incomplete to me. Do ypu think anything is missing? @mig82 .

Reviewer: so i think a `in` should be included in between. 
As shown here : be able to access these properties as they would `in` any other environment
what do you think? @MarkEWaite

variable.

![](docs/images/folder-properties-freestyle-config.png)

> Only then will they inherit properties defined by their parent or ancestor folders —e.g. Running `echo $FOO` in a Shell build step :

Reviewer 2: General guidance from writers is to avoid the Latin abbreviation `e.g.` and replace it with literal 'for example'.  Optional
```suggestion
Only then will they inherit properties defined by their parent or ancestor folders — for example, running `echo $FOO` in a Shell build step :
```

 or `${SCM_URL}` and `${BRANCH_SELECTOR}`:
 or as `${SCM_URL}` and `${BRANCH_SELECTOR}` :

<img src="./pics/freestyle-example-scm-3.png" align="center" alt="freestyle build env" style="width: 80%; height: auto"/><br>
![](docs/images/freestyle-example-scm-3.png)

### Pipeline Jobs
## Pipeline Jobs

Pipeline jobs can use step `withFolderProperties` to access them either inside or outside a node step:
Pipeline jobs can use step `withFolderProperties` to access them :

    withFolderProperties{
        echo("Foo: ${env.FOO}")
    }
**Using folder properties in a pipeline job**

Jenkins deployments using some of the older versions of the
[Structs Plugin](https://wiki.jenkins.io/display/JENKINS/Structs+plugin) will need to do this using the `wrap` meta-step.
In such scenarios you should note that the wrapped syntax must run inside a node step.
``` groovy
withFolderProperties{
    echo("Foo: ${env.FOO}")
}
```

    node {
        wrap([$class: 'ParentFolderBuildWrapper']) {
            echo("Foo: ${env.FOO}")
        }
    }
Jenkins deployments using some of the older versions of the [Structs
Plugin](https://plugins.jenkins.io/structs/) will need
to do this using the `wrap` meta-step :

> ### Declarative Pipeline Jobs

Reviewer 2: Shouldn't the declarative pipeline example be retained?  Or, maybe I missed something and it is now obsolete?

Reviewer 3: Yes, it should be retained.

**Using folder properties in older pipeline jobs**

Declarative pipeline jobs can also use the `options` directive to leverage folder properties as follows:
``` groovy
wrap([$class: 'ParentFolderBuildWrapper']) {
    echo("Foo: ${env.FOO}")
}
```

    pipeline {
        agent any
        options {

>             withFolderProperties()

Reviewer 2: I tried with Jenkins 2.452.4 and recent Pipeline plugin releases and I see unexpected behavior.  The declarative Pipeline directive generator for "options" includes an entry for "withFolderProperties" but the generated example is

```
    options {
        withFolderProperties
    }
```

When I try to run that Pipeline, it fails.

I am able to successfully run a declarative Pipeline that uses the withFolderProperties wrapper inside the step.  Needs more exploring before I'll be ready to submit a proposed documentation update.

> * Javadoc: https://mig82.github.io/folder-properties-plugin/apidocs

Reviewer 2: @mig82 any reason this could not use the standard Jenkins javadoc location instead of using your GitHub pages version?
```suggestion
* Javadoc: https://javadoc.jenkins-ci.org/plugin/folder-properties/
```

Reviewer 3: Note at all. This was my first and only contribution to the Jenkins ecosystem and I obviously missed a thing or two.

# Jenkins Folder Properties Plugin
A Jenkins plugin which allows users with config permission to define
properties for a folder which can then be used by any jobs contained
within it or any of its sub-folders. 

A Jenkins plugin which allows users with config permission to define properties for a folder which can then be used by

> any jobs contained within it or any of its sub-folders.

Reviewer 2: It is usually best for the maintainer if you do not "flow" or change the line breaks on existing text.  The change of line breaks on existing text makes it much more difficult to identify what has changed and what has remained the same.

Reviewer: Thanks a lot for taking your time to review this @MarkEWaite !!. But, I didn't get what you meant by the _flow_ thing. Would you mind explaining a bit more?.

Reviewer 2: > Thanks a lot for taking your time to review this @MarkEWaite !!. But, I didn't get what you meant by the _flow_ thing. Would you mind explaining a bit more?.

Sorry for the confusing phrasing.  When a sentence is word wrapped differently in a pull request than in the original, it appears as a complete line of changes in the GitHub diff output.  That makes the diff larger than necessary and more difficult for the reviewer to review.  I'm from a era when the word ["reflow"](https://www.techwalla.com/articles/what-is-text-reflow) was used to describe that style of text modification.

Reviewer: haha, thanks for clarifying Mark!

> You can find [this plugin's source code on
> Github](https://github.com/jenkinsci/folder-properties-plugin).

Reviewer 2: I'd delete this paragraph since there is already a GitHub link provided from the plugins.jenkins.io page.
```suggestion
```

 

# About this Plugin

The aim here is to remove the need to specify the same properties over
and over again for all the jobs inside a folder.

In structures where two or more folders are nested, any property defined
for a folder will be overridden by any other property of the same name
defined by one of its sub-folders.

## Configuring Folder Properties

Just create a

> [folder](http://localhost:8085/display/JENKINS/CloudBees+Folders+Plugin),

Reviewer 4: ```suggestion
[folder](https://plugins.jenkins.io/cloudbees-folder/),
```

A small suggestion to update the link, as the current link doesn't work.

go to its configuration page and add as many properties as you need
under the `Folder Properties` section.

> ![](docs/images/folder-properties-config.png){height="250"}

Reviewer 4: ```suggestion
![](docs/images/folder-properties-config.png)
```
This property is visible in the file, do we need to add this?

Reviewer: Yes, I guess this property is needed. And I didn't get what you meant by "_visible_" in file. Can you explain more?

Reviewer 4: Sure, so if you go to your copy of the README.md file, you'd be able to see "{height="250"}" displayed just below your image. We might want to remove it from there.

Reviewer: Yup, understood. Thanks for pointing it out @dheerajodha !!. I'll fix it asap.

## Freestyle Jobs

Freestyle jobs must opt into the `Folder Properties` build wrapper from
the `Build Environment` section of their configuration page in order to
be able to access these properties as they would any other environment
variable.

![](docs/images/folder-properties-freestyle-config.png)

## Pipeline Jobs

Pipeline jobs can use step `withFolderProperties` to access them:

**Using folder properties in a pipeline job**

``` syntaxhighlighter-pre
withFolderProperties{
    echo("Foo: ${env.FOO}")
}
```

Jenkins deployments using some of the older versions of the [Structs

> Plugin](http://localhost:8085/display/JENKINS/Structs+plugin) will need

Reviewer 4: ```suggestion
Plugin](https://plugins.jenkins.io/structs/) will need
```

Reviewer: thanks!!

to do this using the `wrap` meta-step:

**Using folder properties in older pipeline jobs**

> ``` syntaxhighlighter-pre

Reviewer 2: We usually use groovy syntax highlighter for Pipeline steps.
```suggestion
``` groovy
```

Reviewer: fixed. coming on next commit.

wrap([$class: 'ParentFolderBuildWrapper']) {
    echo("Foo: ${env.FOO}")
}
```

## Job DSL

In Job DSL scripts you can define folder properties like so:

**Job DSL example**

> ``` syntaxhighlighter-pre

Reviewer 2: The groovy syntax highlighter seems more likely to do what we want
```suggestion
``` groovy
```
